src/types.ts

```diff
@@ -268,7 +268,7 @@ export type RefSchemaType =
  * Reference another Model
  */
 // export type Ref<R, T extends RefType = mongoose.Types.ObjectId> = R | T; // old type, kept for easy revert
-export type Ref<R, T extends RefType = R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId> = R | T;
+export type Ref<R, T extends RefType = (R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId) | undefined> = R | T;
```

Comment:
what does this extra `| undefined` do here?

Reply:
Some more context here:
https://github.com/microsoft/TypeScript/issues/38323#issuecomment-626183067

Assume this code:

```ts
declare const ref: Ref<SomeOtherModel> | undefined;

if (isDocument(ref)) {
```

Original `Ref` type:
```ts
export type Ref<R, T extends RefType = R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId> = R | T;
```

If `ref` is `undefined`, then `Ref<R, T>` works with two `undefined` values for both `R` and `T`, but `R` would not work with `undefined` because `R extends { _id?: RefType } ? NonNullable<R['_id']> : mongoose.Types.ObjectId` cannot ever be `undefined` by that definition.

So it just adds it as a possible value for `R`, which is valid for the case I mentioned above.